lib/ecto/migrator.ex

```diff
@@ -423,29 +436,173 @@ defmodule Ecto.Migrator do
+    |> collect_migrations(directories, opts)
     |> Enum.sort_by(fn {_, version, _} -> version end)
   end
 
-  defp collect_migrations(versions, migration_source) do
+  defp run_smart(repo, migration_source, opts = [smart: true]) do
+    migration_source = List.wrap(migration_source)
+
+    pending =
+      lock_for_migrations true, repo, opts, fn _config, versions ->
+        versions_with_hash =
+          versions
+          |> Enum.map(fn [version | _] -> version end)
+          |> pending_in_direction(migration_source, :smart)
+          |> Enum.map(&load_migration!/1)
+          |> Enum.map(&hash_migration_source/1)
+          |> Enum.map(fn {version, _, hash} -> {version, hash} end)
+          |> Enum.into(%{})
+
+        versions
+        |> Enum.filter(fn [_, migration_script] -> migration_script != nil end)
+        |> Enum.map(&load_migration!/1)
+        |> Enum.map(&hash_migration_source/1)
+        |> Enum.sort(fn {left, _, _}, {right, _, _} -> left < right end)
+        |> needs_down?(versions_with_hash, :no_change)
+        |> Enum.reverse
+      end
+
+    # The lock above already created the table, so we can now skip it.
+    opts = Keyword.put(opts, :skip_table_creation, true)
+
+    ensure_no_duplication!(pending)
+    migrate(pending, :smart_down, repo, opts)
+    run(repo, migration_source, :up, [all: true])
+  end
+
+  defp hash_migration_source([]), do: []
+
+  defp hash_migration_source([version, migration_source]) do
+    {version, migration_source, hash_migration_source(migration_source)}
+  end
+
+  defp hash_migration_source({version, module, migration_source}) do
+    {version, module, hash_migration_source(migration_source)}
+  end
+
+  defp hash_migration_source(migration_source) when is_binary(migration_source) do
+    :crypto.hash(:sha256, migration_source)
+    |> Base.encode16()
+    |> String.downcase()
+  end
+
+  defp needs_down?([{version, mod, hash} | migrations_on_file], complete_migrations, :no_change) when is_map(complete_migrations) do
+    if Map.get(complete_migrations, version) == hash do
+      needs_down?(migrations_on_file, complete_migrations, :no_change)
+    else
+      [{version, mod, ""}] ++ needs_down?(migrations_on_file, :needs_down)
+    end
+  end
+
+  defp needs_down?([], _, :no_change), do: []
+
+  defp needs_down?([{version, mod, _} | migrations_on_file], :needs_down) do
+    [{version, mod, ""}] ++ needs_down?(migrations_on_file, :needs_down)
+  end
+
+  defp needs_down?([], :needs_down), do: []
+
+  defp smart_migrations_confirm(repo) do
+    opts = [smart: true]
+    paths = Mix.EctoSQL.ensure_migrations_paths(repo, opts)
+
+    IO.puts(
+      """
+
+      Repo: #{inspect(repo)}
+
+       Status    Migration ID    Migration Name
+      --------------------------------------------------
+      """ <>
+      Enum.map_join(migrations(repo, paths, opts), "\n", fn {status, number, description} ->
+       "  #{format(status, 10)}#{format(number, 16)}#{description}"
```

Comment:
Would it be possible here to differentiate a `diff` from a file without a migration script? The script generated this output for me:
```  
  diff      20201007210911  Elixir.FunraiseFeatures.Repo.Migrations.Sites
  diff      20201119211250  Elixir.FunraiseFeatures.Repo.Migrations.AddStripeTerminalFeature
  diff      20210209162752  Elixir.FunraiseFeatures.Repo.Migrations.TestUp1
  up        20210209163152  Elixir.FunraiseFeatures.Repo.Migrations.TestUp2
```

The `Sites` and `AddStripeTerminalFeature` were created before I had smart evolutions, so they do not have migration scripts. `TestUp1` and `TestUp2` do indeed have migration scripts, they were both run up, and then I changed `TestUp1`. I would expect that the smart migration will then run:
* `TestUp2` - down
* `TestUp1` - down
* `TestUp1` - up
* `TestUp2` - up

In my testing, that did happen, but it is unclear in the output that it will stop running ups and downs at `TestUp1` - it would be nice if it told me that it was going to ignore ups/downs for `AddStripeTerminalFeature` and anything before it. It would also be nice if it could calculate and show me that it *will* run downs for `TestUp2`, but I'm not sure how easy that is and how worth it it will be